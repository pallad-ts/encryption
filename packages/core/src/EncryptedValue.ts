import {Either, Validation} from "monet";
import * as is from 'predicates'

const _isBase64 = require('is-base64');

const HEX_PATTERN = /^([0-9a-f]+)$/
function isHex(value: string) {
    return HEX_PATTERN.test(value);
}

function isBase64(value: string) {
    return _isBase64(value, {
        allowEmpty: false
    });
}

export class EncryptedValue {
    constructor(readonly iv: Buffer, readonly encrypted: Buffer) {
        if (iv.length === 0 || encrypted.length === 0) {
            throw new Error('IV or Encrypted value are empty');
        }

        Object.freeze(this);
    }

    toString(encoding: EncryptedValue.Encoding = 'hex') {
        return `${this.iv.toString(encoding)}:${this.encrypted.toString(encoding)}`;
    }

    asTuple(): [Buffer, Buffer] {
        return [
            this.iv,
            this.encrypted
        ];
    }

    static fromString(str: string, encoding: EncryptedValue.Encoding = 'hex'): Validation<string, EncryptedValue> {
        const result = str.split(':', 2);

        if (result.length !== 2) {
            return Validation.Fail('Malformed encrypted string');
        }

        if (is.blank(result[0]) || is.blank(result[1])) {
            return Validation.Fail('IV or Encrypted value are empty');
        }

        const isValidIv = encoding === 'hex' ? isHex(result[0]) : isBase64(result[0]);
        const isValidEncrypted = encoding === 'hex' ? isHex(result[1]) : isBase64(result[1]);

        if (!isValidIv || !isValidEncrypted) {
            return Validation.Fail(`Invalid value for encoding: ${encoding}`);
        }

        return Validation.Success(
            EncryptedValue.fromStringFormatted(result[0], result[1], encoding).success()
        );
    }

    static fromStringFormatted(iv: string, encrypted: string, encoding: EncryptedValue.Encoding = 'hex'): Validation<string, EncryptedValue> {
        return Either.fromTry(() => {
            return new EncryptedValue(
                Buffer.from(iv, encoding),
                Buffer.from(encrypted, encoding)
            )
        })
            .leftMap(x => x.message)
            .toValidation();
    }
}

export namespace EncryptedValue {
    export type Encoding = 'base64' | 'hex';
}