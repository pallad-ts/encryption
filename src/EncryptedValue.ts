import {Either, Validation} from "monet";
import * as is from 'predicates'

export class EncryptedValue {
    constructor(readonly iv: Buffer, readonly encrypted: Buffer) {
        Object.freeze(this);
    }

    toString(encoding: EncryptedValue.Encoding = 'hex') {
        return `${this.iv.toString(encoding)}:${this.encrypted.toString(encoding)}`;
    }

    asTuple(): [Buffer, Buffer] {
        return [
            this.iv,
            this.encrypted
        ];
    }

    static fromString(str: string, encoding: EncryptedValue.Encoding = 'hex'): Validation<string, EncryptedValue> {
        const result = str.split(':', 2);

        if (result.length !== 2) {
            return Validation.Fail('Malformed encrypted string');
        }

        if (is.blank(result[0]) || is.blank(result[1])) {
            return Validation.Fail('IV or Encrypted value are empty');
        }

        return Validation.Success(
            EncryptedValue.fromStringFormatted(result[0], result[1], encoding).success()
        );
    }

    static fromStringFormatted(iv: string, encrypted: string, encoding: EncryptedValue.Encoding = 'hex'): Validation<string, EncryptedValue> {
        return Either.fromTry(() => {
            return new EncryptedValue(
                Buffer.from(iv, encoding),
                Buffer.from(encrypted, encoding)
            )
        })
            .leftMap(x => x.message)
            .toValidation();
    }
}

export namespace EncryptedValue {
    export type Encoding = 'base64' | 'hex';
}