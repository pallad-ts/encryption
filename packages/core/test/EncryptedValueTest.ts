import {EncryptedValue} from "@src/EncryptedValue";
import {randomBytes} from "crypto";

describe('EncryptedValue', () => {

    describe('creating', () => {
        describe('from string', () => {
            it('success', () => {
                const iv = randomBytes(10);
                const encrypted = randomBytes(30);
                const input = `${iv.toString('hex')}:${encrypted.toString('hex')}`;
                const value = EncryptedValue.fromString(input).success();
                expect(value.toString())
                    .toEqual(input);
                expect(value)
                    .toEqual(new EncryptedValue(iv, encrypted));
            });

            describe('fail', () => {
                it('malformed', () => {
                    const result = EncryptedValue.fromString('invalid');

                    expect(result.isFail())
                        .toBe(true);

                    expect(result.fail())
                        .toMatchSnapshot();
                });

                it.each([
                    [':'],
                    [':0000'],
                    ['   :0000'],
                    ['0000:'],
                    ['0000:   ']
                ])('One of value is blank or empty: %s', value => {
                    const result = EncryptedValue.fromString(value);
                    expect(result.isFail())
                        .toBe(true);

                    expect(result.fail())
                        .toMatchSnapshot();
                });

                it.each([
                    ['10:hg'],
                    ['hg:10']
                ])('Invalid hex: %s', value => {
                    const result = EncryptedValue.fromString(value);

                    expect(result.isFail())
                        .toBe(true);

                    expect(result.fail())
                        .toMatchSnapshot();
                });

                it.each([
                    ['uuLMhh==:dfasdfr342'],
                    ['dfasdfr342:uuLMhh==']
                ])('Invalid base64: %s', value => {
                    const result = EncryptedValue.fromString(value, 'base64');

                    expect(result.isFail())
                        .toBe(true);

                    expect(result.fail())
                        .toMatchSnapshot();
                });
            });
        });

        describe('from formatted string', () => {
            it.each<[EncryptedValue.Encoding]>([
                ['hex'],
                ['base64']
            ])('encoding: %s', (encoding) => {
                const iv = randomBytes(10);
                const encrypted = randomBytes(50);
                const result = EncryptedValue.fromStringFormatted(
                    iv.toString(encoding),
                    encrypted.toString(encoding),
                    encoding
                );
                expect(result.success())
                    .toMatchObject({
                        iv: iv,
                        encrypted: encrypted
                    })
            });

            it('fail - invalid value for encoding', () => {
                const result = EncryptedValue.fromStringFormatted('jnkm', randomBytes(2).toString('hex'));

                expect(result.isFail())
                    .toBe(true);

                expect(result.fail())
                    .toMatchSnapshot();
            });
        });
    });


});