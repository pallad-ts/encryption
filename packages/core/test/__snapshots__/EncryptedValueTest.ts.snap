// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EncryptedValue creating from formatted string fail - invalid value for encoding 1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail Invalid base64: dfasdfr342:uuLMhh== 1`] = `"Invalid value for encoding: base64"`;

exports[`EncryptedValue creating from string fail Invalid base64: uuLMhh==:dfasdfr342 1`] = `"Invalid value for encoding: base64"`;

exports[`EncryptedValue creating from string fail Invalid hex: 10:hg 1`] = `"Invalid value for encoding: hex"`;

exports[`EncryptedValue creating from string fail Invalid hex: hg:10 1`] = `"Invalid value for encoding: hex"`;

exports[`EncryptedValue creating from string fail One of value is blank or empty:    :0000 1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail One of value is blank or empty: : 1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail One of value is blank or empty: :0000 1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail One of value is blank or empty: 0000:    1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail One of value is blank or empty: 0000: 1`] = `"IV or Encrypted value are empty"`;

exports[`EncryptedValue creating from string fail malformed 1`] = `"Malformed encrypted string"`;
